Use bloom filters to reduce tree traversal

emjin opened the pull request:
Closes #2352

When deciding whether or not to match a subtree, we can definitely choose not to if essential parts of the pattern are not present. It is cheap and space efficient to first hash these essential parts using a bloom filter, which will guarantee a "no" answer if it is not present. Then, when matching the pattern against a node, we only traverse further if all essential parts of the pattern are present in that node.

For the current implementation:
- the essential parts are non-metavariable idents and string literals
- a bloom filter is stored at every statement node

The initial bloom annotation is worst case O(sn), where s is the number of strings and n is the number of statement nodes. This is because each string for a given statement is added to that statement's bloom filter and each of its parents'. Note that this is a *very* loose bound

Later computations will be O(s_p), where s_p is the number of strings in the pattern.

Test plan: make test (correctness should still hold), benchmarking to see if it is faster. Will also need to tweak the parameters passed into Bloomf.create

This implementation makes use of the bloomf library in opam

emjin committed:
Use bloom filters to reduce tree traversal

Closes #2352

When deciding whether or not to match a subtree, we can definitely choose not to if essential parts of the pattern are not present. It is cheap and space efficient to first hash these essential parts using a bloom filter, which will guarantee a "no" answer if it is not present. Then, when matching the pattern against a node, we only traverse further if all essential parts of the pattern are present in that node.

For the current implementation:
- the essential parts are non-metavariable idents and string literals
- a bloom filter is stored at every statement node

The initial bloom annotation is worst case O(sn), where s is the number of strings and n is the number of statement nodes. This is because each string for a given statement is added to that statement's bloom filter and each of its parents'.

Later computations will be O(s_p), where s_p is the number of strings in the pattern.

Test plan: make test (correctness should still hold), benchmarking to see if it is faster. Will also need to tweak the parameters passed into Bloomf.create

This implementation makes use of the bloomf library in opam

emjin committed:
Create a list of strings in the pattern rather than a bloom filter for them

emjin committed:
Renamed in semgrep_generic and fixed infinite loop

Changed bf to pattern_strs where it originally represented the bloom filter

Had infinite loop because the visitor visits the top level statement. Hacky fix--just always skip the first statement reached

emjin committed:
Ignore string literal semgrep constructs (..., regexp)

emjin committed:
Traverse id_info in visitor

Added a hook for id_info. To maintain the same default behavior, v_id_info enters the refs, but the default hook kid_info does not. Bloom_annotation overrides that by repalacing the hook with one that traverses

emjin committed:
Used constant propagation information and added ident special cases

Add hook to constness to match literals. For idents, it was necessary to exclude strings starting with "$..." and regexps. This is currently done poorly.

emjin committed:
Removed print statements

emjin committed:
Resolvd conflict

mjambon commented:
@pad can we eliminate this semgrep rule:
```
severity:warning rule:ocaml.lang.best-practice.string.ocamllint-str-first-chars: Use instead `Str.first_chars`
82:  AST_generic_.is_metavar_name str || (String.length str > 4 && (String.sub str 0 4) = "$...") || is_regexp_string str
```
`Str.first_chars "abc" 4` is the same as `String.sub "abc" 0 4`, both raise the same exception. And `Str` requires the str library.

aryx commented:
I like the semgrep check. It's easier to read, but I let @emjin decide if she thinks it's a useless rule.
You can update the semgrep-rules repo to remove the rule if you want.

aryx commented on semgrep-core/Analyzing/Bloom_annotation.ml:
```diff
@@ -41,9 +44,44 @@ module B = Bloom_filter
+*)
+
+let push v (l : 'a list ref) = Common.push v l
+;;
```
you don't need those ;;
They are need only in ocamltop (the OCaml REPL).

aryx commented on semgrep-core/Analyzing/Bloom_annotation.ml:
```diff
@@ -53,23 +91,94 @@ module B = Bloom_filter
+       | _ -> k x
+      )
+    );
+    V.kid_info = (fun (k, _) x -> k x);
```
maybe a comment here reminding that the default behavior of kid_info is to not call the continuation? which
is why we redefine here to call the continuation so we can also index propagated constants inside id_constness
(that would not be indexed by the default visitor).

aryx commented on semgrep-core/Analyzing/Bloom_annotation.ml:
```diff
@@ -53,23 +91,94 @@ module B = Bloom_filter
+      if !top_level then (top_level := false; k x)
+      else
+        (* For any other statement, recurse to add the filter *)
+        (let strs = statement_strings x in
```
I usually prefer begin/end rather than () to mark a block of instructions.

aryx commented on semgrep-core/Analyzing/Bloom_annotation.ml:
```diff
@@ -53,23 +91,94 @@ module B = Bloom_filter
+    V.kid_info = (fun (k, _) x -> k x);
+    V.kstmt = (fun (k, _) x ->
+      (* First statement visited is the current statement *)
+      if !top_level then (top_level := false; k x)
```
why do we do something special for the top statement?
why we don't set its bloom filter too?

emjin replied:
This is because of how I visit so that I can get all the strings in a statement (including its substatements). I use the visitor until I come to a statement S. Then, I stop that visitor, and recursively call the statement_strings function, which starts a ref for a list of strings and creates a new visitor for S. However, S was already visited by the previous visitor, so it should not be revisited. I couldn't quite figure out how to do this with only one visitor call, though I'm afraid this will use more memory.

aryx commented on semgrep-core/Analyzing/Bloom_annotation.ml:
```diff
@@ -53,23 +91,94 @@ module B = Bloom_filter
+let list_of_pattern_strings any =
+  let res = ref [] in
+  let visitor = V.mk_visitor {
+    V.default_visitor with
```
looks like duplicated code with above. Maybe you can factorize.

emjin replied:
I wasn't sure what the best way to do this was. The code is very similar in some ways, but statement_strings is a recursive function, called recursively at each statement node, whereas list_of_pattern_strings visits the entire AST. I could make functions to produce kident and kexpr which take in a ref, I just wasn't sure if it would end up making the code less readable. If I start making kexpr more complicated, I'll factorize it.

aryx replied:
ok, sounds fair.

aryx commented on semgrep-core/Analyzing/Bloom_annotation.ml:
```diff
@@ -53,23 +91,94 @@ module B = Bloom_filter
+    V.default_visitor with
+    V.kstmt = (fun (_k, _) x ->
+      let bf = B.create () in
+      let ids = statement_strings x in
```
This seems like duplicated with the statement_strings code, why not factorize? (which would also
make useless the is_toplevel variable).

emjin replied:
See earlier comment (the need to union all the child statements' strings and add it to the statement's bloom filter). There may be a better way of doing this (perhaps passing in a ref list of bloom filters?), but I'm not sure

aryx commented on semgrep-core/Core/AST/Bloom_filter.ml:
```diff
@@ -15,29 +16,45 @@
 let create () =
-  { added = false }
+  { added = false;
+    filter = B.create ~error_rate:0.01 10 }
```
maybe some quick explanation of what means the parameter and what is the intuition behind  those numbers?

emjin committed:
Also filter pattern in generic_vs_generic

Constant propagation test case in java has more cases so that I could measure speedup

aryx commented on semgrep-core/Matching/Generic_vs_generic.ml:
```diff
@@ -1643,143 +1654,152 @@ and m_list__m_stmt_uncached ~flattened (xsa: A.stmt list) (xsb: A.stmt list) =
-  | A.Assert _, _
-  | A.OtherStmt _, _ | A.OtherStmtWithStmt _, _ | A.WithUsingResource _, _
-    -> fail ()
+  let pattern_strs = Bloom_annotation.list_of_pattern_strings (S a) in
```
No! I don't think it's a good idea to do this each time in m_stmt.
Either we also pre-compute also on the pattern the bloom filter, and set it in the s_bf field of the pattern,
or we leave the logic outside Generic_vs_generic. Generic_vs_generic is already quite complicated
so I would really prefer to have the optimizations done outside, e.g., in Semgrep_generic.ml as it was before.

emjin replied:
This was an experiment---I was planning on asking you for advice on Monday, but I'm not getting any speedup on stmt lists because it doesn't check the pattern against a bloom filter first. See my reply to Martin in eng.

emjin committed:
Reverted Generic_vs_generic changes

Use of bloom filter check in generic_vs_generic did not make a difference

emjin committed:
Returned must_analyze_statement_bloom... func to semgrep_generic

emjin committed:
Removed print statements

emjin committed:
Addressed comments

emjin committed:
Use Str.first_chars str 4 instead of String.sub

TODO: use a regex

emjin commented:
> I like the semgrep check. It's easier to read, but I let @emjin decide if she thinks it's a useless rule.
> You can update the semgrep-rules repo to remove the rule if you want.

I left the rule and changed my code, but there's probably a way to do what I want with regexs (precede with (...)*?)

emjin committed:
Profile bloom annotating

aryx commented:
Why the CI failures? Can you fix that so we can merge it after?

emjin committed:
Only check for metavariables/ellipses in the pattern

emjin committed:
Merge branch 'develop' into EJ-2352-bloom-filter

emjin committed:
Added -bloom_filter to turn on the bloom filter, default off

emjin committed:
Added a comment, removed a print statement

aryx commented on semgrep-core/CLI/Main.ml:
```diff
@@ -1288,6 +1288,8 @@ let options () =
     "-filter_irrelevant_rules", Arg.Set Flag.filter_irrelevant_rules,
     " filter rules not containing any strings in target file";
+    "-bloom_filter", Arg.Set Flag.use_bloom_filter,
+    "-no_bloom_filter";
```
This seems wrong, this string is supposed to comment what -bloom_filter means.
